### DISCUSSION TITLE
The way to use the second GPU for the JointTextEncoder

### DISCUSSION THREAD

**Discussion by Juqowel** [lllyasviel/stable-diffusion-webui-forge#1591] [Ideas]
I'm already did simple, but VERY BAD implementation for youself. (for 4060ti(16gb) and 3060(12gb))
It looks like:

![Снимок](https://github.com/user-attachments/assets/e210c068-400b-4cef-ae38-123ba1a3ec1f)

Saves a lot of time.

First load singe gpu:
![first - single](https://github.com/user-attachments/assets/cecedb43-9e65-499e-a33e-bf36c1ecef5e)
Second+ load with prompt changed, single gpu:
![second - single](https://github.com/user-attachments/assets/10474c28-a0b8-49e5-a061-e2df57e37275)
First load dual gpu:
![first - dual](https://github.com/user-attachments/assets/4b2ce90c-d5f7-477b-9656-a98624a9852b)
Second+ load with prompt changed, dual gpu:
![second - dual](https://github.com/user-attachments/assets/e601c1cc-ac90-4138-a946-d19df4ad5d98)

I think it would be great if this feature was added for everyone and with GOOD implementation.

**Comment by lllyasviel** [comment]
You can do things in a “Forge” way:

Just create a folder and put this like `extensions/what_ever_name_you_like/scripts/whatever_you_like.py` 

```python
import torch
import gradio as gr
from modules import scripts


class T5onOtherDevice(scripts.Script):
    def title(self):
        return "T5 on Other Device"

    def show(self, is_img2img):
        return scripts.AlwaysVisible

    def ui(self, *args, **kwargs):
        with gr.Accordion(open=False, label=self.title()):
            enabled = gr.Checkbox(label='Enabled', value=False)
        return enabled

    def process(self, p, *script_args, **kwargs):
        self.enabled = script_args
        if not self.enabled:
            # Todo； add some codes to revert the change
            return
        p.sd_model.forge_objects.clip.patcher.load_device=torch.device('cuda:1')
        return
```

But this code may need a bit more tests

**Reply by Juqowel** [reply]
Script done. Not ideal, but for me works perfect:

```
import torch
import gradio as gr
from modules import scripts
from backend import memory_management

lowvram_available = memory_management.lowvram_available

class T5onOtherDevice(scripts.Script):

    def title(self):
        return "T5 on Other Device"

    def show(self, is_img2img):
        return scripts.AlwaysVisible

    def ui(self, *args, **kwargs):
        gpus = [f'cuda:{i}' for i in range(0, torch.cuda.device_count())]
        if len(gpus) < 2:
             return []
        gpus.remove(f'{memory_management.get_torch_device()}')
        with gr.Accordion(open=False, label=self.title()):
            enabled = gr.Checkbox(label='Enabled', value=False)
            choise = gr.Radio(label='Select GPU', choices=gpus, value=gpus[0])
        return [choise, enabled]

    def process(self, p, *script_args, **kwargs):
        if not script_args[0]:
            return
        if not script_args[1]:
            p.sd_model.forge_objects.clip.patcher.load_device=torch.device(f'{memory_management.get_torch_device()}')
            memory_management.lowvram_available = lowvram_available
            return
        memory_management.lowvram_available = False
        p.sd_model.forge_objects.clip.patcher.load_device=torch.device(f'{script_args[0]}')
        return
```

Will you add it to the built-in? If not, I'll just save it [here](https://github.com/Juqowel/GPU_For_T5).

**Reply by Iory1998** [reply]
@Juqowel I was referred to this comment by @lllyasviel (https://github.com/lllyasviel/stable-diffusion-webui-forge/discussions/1753).  
Could you please update the script so that it force the text-encoders be loaded in RAM instead like the force/set Device node in comfyUI?
![image](https://github.com/user-attachments/assets/1370d359-468c-48de-90e7-8afa5eaa6bc1)

**Reply by Juqowel** [reply]
@Iory1998 Ready. Test it, please.

**Reply by Juqowel** [reply]
@lllyasviel
Maybe **forge** will load t5 to RAM by default (and _before_ t5 - Kmodel - to the main GPU)?
Seems like this is faster then second GPU.
2nd GPU:
![_gpu](https://github.com/user-attachments/assets/87fa49fe-3d02-429a-ace3-89cddec4865b)
CPU:
![_cpu](https://github.com/user-attachments/assets/7911092f-6221-4365-93af-5c4f94b8fca5)

And one more problem - t5 gguf doesn't load. [ERROR: Expected all tensors to be on the same device](https://github.com/Juqowel/GPU_For_T5/issues/1#issue-2512755556)

**Reply by Iory1998** [reply]
> @Iory1998 Ready. Test it, please.

@Juqowel, I've been trying the extension for 45 min now with different Flux model quants, T5 quants, and LoRAs. The first time I activated your extension, I got an error stating (I'm paraphrasing) that the models are split between 2 places (between GPU and CPU) and the app closed. I couldn't replicate the error.

If you use the Full FP model without your extension, you would see that the text-encoders get loaded first into the Vram, then the memory management tries to empty the VRAM, which leads to my 32GB RAM get 100% used. This lasts for a few minutes before the VRAM is emptied, and the fp16 model is being loaded instead of the text-encoders. 

When your extension is being activated, I noticed the flux models is being loaded first into VRAM. With smaller quants, that leads to better RAM management and the speed is up.

I wish that Forge priorities loading the Flux models first with LoRAs and then see if there is enough VRAM capacity to add the text-encoders. I notice similar behavior with ComfyUI, so maybe this is technical and there is no way around it.

I'll come back when I make more tests.

P.S. I have an RTX3090 and 32GB of DDR4 RAM.

**Reply by Iory1998** [reply]
I managed to face the runtimeError: xpected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)
Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)

Here is the full log:

_Python 3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]
Version: f2.0.1v1.10.1-previous-525-gefe6fed4
Commit hash: efe6fed4994e6d29627d5172aecde71e7e496255
I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\extensions-builtin\forge_legacy_preprocessors\install.py:2: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\extensions-builtin\sd_forge_controlnet\install.py:2: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
Launching Web UI with arguments: --theme dark --gradio-allowed-path 'I:\StabilityMatrix-win-x64\Data\Images'
Total VRAM 24576 MB, total RAM 32317 MB
pytorch version: 2.3.1+cu121
Set vram state to: NORMAL_VRAM
Device: cuda:0 NVIDIA GeForce RTX 3090 : native
Hint: your device supports --cuda-malloc for potential speed improvements.
VAE dtype preferences: [torch.bfloat16, torch.float32] -> torch.bfloat16
CUDA Using Stream: False
Using pytorch cross attention
Using pytorch attention for VAE
ControlNet preprocessor location: I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\models\ControlNetPreprocessor
Tag Autocomplete: Could not locate model-keyword extension, Lora trigger word completion will be limited to those added through the extra networks menu.
2024-09-09 10:58:51,420 - ControlNet - INFO - ControlNet UI callback registered.
Model selected: {'checkpoint_info': {'filename': 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\Stable-diffusion\\Flux\\flux1Dev_v10.safetensors', 'hash': 'b04b3ba1'}, 'additional_modules': ['I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\VAE\\Flux\\VAE_Flux_Dev.safetensors', 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\text_encoder\\clip_l.safetensors', 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\text_encoder\\t5xxl_fp16.safetensors'], 'unet_storage_dtype': None}
Using online LoRAs in FP16: True
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 15.1s (prepare environment: 3.6s, import torch: 6.0s, initialize shared: 0.2s, other imports: 0.5s, load scripts: 2.1s, initialize extra networks: 0.1s, create ui: 1.8s, gradio launch: 0.8s).
Environment vars changed: {'stream': False, 'inference_memory': 6135.0, 'pin_shared_memory': True}
[GPU Setting] You will use 75.04% GPU memory (18440.00 MB) to load weights, and use 24.96% GPU memory (6135.00 MB) to do matrix computation.
Model selected: {'checkpoint_info': {'filename': 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\Stable-diffusion\\Flux\\flux1Dev_v10.safetensors', 'hash': 'b04b3ba1'}, 'additional_modules': ['I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\VAE\\Flux\\VAE_Flux_Dev.safetensors', 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\text_encoder\\clip_l.safetensors'], 'unet_storage_dtype': None}
Using online LoRAs in FP16: True
Model selected: {'checkpoint_info': {'filename': 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\Stable-diffusion\\Flux\\flux1Dev_v10.safetensors', 'hash': 'b04b3ba1'}, 'additional_modules': ['I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\VAE\\Flux\\VAE_Flux_Dev.safetensors', 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\text_encoder\\clip_l.safetensors', 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\text_encoder\\t5-v1_1-xxl-encoder-Q8_0.gguf'], 'unet_storage_dtype': None}
Using online LoRAs in FP16: True
Environment vars changed: {'stream': False, 'inference_memory': 6135.0, 'pin_shared_memory': False}
[GPU Setting] You will use 75.04% GPU memory (18440.00 MB) to load weights, and use 24.96% GPU memory (6135.00 MB) to do matrix computation.
Environment vars changed: {'stream': False, 'inference_memory': 6135.0, 'pin_shared_memory': True}
[GPU Setting] You will use 75.04% GPU memory (18440.00 MB) to load weights, and use 24.96% GPU memory (6135.00 MB) to do matrix computation.
T5 on Other Device is enabled for txt2img and cpu
Loading Model: {'checkpoint_info': {'filename': 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\Stable-diffusion\\Flux\\flux1Dev_v10.safetensors', 'hash': 'b04b3ba1'}, 'additional_modules': ['I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\VAE\\Flux\\VAE_Flux_Dev.safetensors', 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\text_encoder\\clip_l.safetensors', 'I:\\StabilityMatrix-win-x64\\Data\\Packages\\stable-diffusion-webui-forge\\models\\text_encoder\\t5-v1_1-xxl-encoder-Q8_0.gguf'], 'unet_storage_dtype': None}
[Unload] Trying to free all memory for cuda:0 with 0 models keep loaded ... Done.
StateDict Keys: {'transformer': 780, 'vae': 244, 'text_encoder': 196, 'text_encoder_2': 219, 'ignore': 0}
Using Detected T5 Data Type: gguf
Using pre-quant state dict!
GGUF state dict: {'Q8_0': 168}
Working with z of shape (1, 16, 32, 32) = 16384 dimensions.
K-Model Created: {'storage_dtype': torch.bfloat16, 'computation_dtype': torch.bfloat16}
Model loaded in 27.2s (unload existing model: 0.2s, forge model load: 27.0s).
[LORA] Loaded I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\models\Lora\Flux\Flux Enhancers\Flux_Skin_Texture.safetensors for KModel-UNet with 304 keys at weight 1.0 (skipped 0 keys) with on_the_fly = False
[LORA] Loaded I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\models\Lora\Flux\Flux Styles\flux_realism_lora.safetensors for KModel-UNet with 76 keys at weight 0.75 (skipped 0 keys) with on_the_fly = False
Skipping unconditional conditioning when CFG = 1. Negative Prompts are ignored.
All loaded to GPU.
Moving model(s) has taken 2.68 seconds
Traceback (most recent call last):
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules_forge\main_thread.py", line 30, in work
    self.result = self.func(*self.args, **self.kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\txt2img.py", line 121, in txt2img_function
    processed = processing.process_images(p)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\processing.py", line 816, in process_images
    res = process_images_inner(p)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\processing.py", line 929, in process_images_inner
    p.setup_conds()
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\processing.py", line 1519, in setup_conds
    super().setup_conds()
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\processing.py", line 501, in setup_conds
    self.c = self.get_conds_with_caching(prompt_parser.get_multicond_learned_conditioning, prompts, total_steps, [self.cached_c], self.extra_network_data)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\processing.py", line 470, in get_conds_with_caching
    cache[1] = function(shared.sd_model, required_prompts, steps, hires_steps, shared.opts.use_old_scheduling)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\prompt_parser.py", line 262, in get_multicond_learned_conditioning
    learned_conditioning = get_learned_conditioning(model, prompt_flat_list, steps, hires_steps, use_old_scheduling)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\modules\prompt_parser.py", line 189, in get_learned_conditioning
    conds = model.get_learned_conditioning(texts)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\diffusion_engine\flux.py", line 79, in get_learned_conditioning
    cond_t5 = self.text_processing_engine_t5(prompt)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\text_processing\t5_engine.py", line 123, in __call__
    z = self.process_tokens([tokens], [multipliers])[0]
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\text_processing\t5_engine.py", line 134, in process_tokens
    z = self.encode_with_transformers(tokens)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\text_processing\t5_engine.py", line 60, in encode_with_transformers
    z = self.text_encoder(
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\nn\t5.py", line 205, in forward
    return self.encoder(x, *args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\nn\t5.py", line 186, in forward
    x, past_bias = l(x, mask, past_bias)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\nn\t5.py", line 162, in forward
    x, past_bias = self.layer[0](x, mask, past_bias)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\nn\t5.py", line 149, in forward
    output, past_bias = self.SelfAttention(self.layer_norm(x), mask=mask, past_bias=past_bias)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\nn\t5.py", line 130, in forward
    past_bias = self.compute_bias(x.shape[1], x.shape[1], x.device, x.dtype)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\nn\t5.py", line 121, in compute_bias
    values = self.relative_attention_bias(relative_position_bucket).to(dtype)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\backend\operations.py", line 343, in forward
    return super().forward(x)
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\modules\sparse.py", line 163, in forward
    return F.embedding(
  File "I:\StabilityMatrix-win-x64\Data\Packages\stable-diffusion-webui-forge\venv\lib\site-packages\torch\nn\functional.py", line 2264, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)
Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument index in method wrapper_CUDA__index_select)_

**Reply by Iory1998** [reply]
> @lllyasviel Maybe **forge** will load t5 to RAM by default (and _before_ t5 - Kmodel - to the main GPU)? Seems like this is faster then second GPU. 2nd GPU: ![_gpu](https://private-user-images.githubusercontent.com/130811765/365574252-87fa49fe-3d02-429a-ace3-89cddec4865b.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU4NzY0MTksIm5iZiI6MTcyNTg3NjExOSwicGF0aCI6Ii8xMzA4MTE3NjUvMzY1NTc0MjUyLTg3ZmE0OWZlLTNkMDItNDI5YS1hY2UzLTg5Y2RkZWM0ODY1Yi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwOVQxMDAxNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZmUyMjVhNjJhZTc0YWFhZjQ4ZDM1NjI0Yzk3YjBjOGUxOWUwMzAwMWQ5MjU3ODNkNjFhNmUyY2I0NzA2NTJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Y5NkE-t_UKwl3Xfv4X-JZI_cQVZQvDf4ZJqdNPF_nH0) CPU: ![_cpu](https://private-user-images.githubusercontent.com/130811765/365574279-7911092f-6221-4365-93af-5c4f94b8fca5.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU4NzY0MTksIm5iZiI6MTcyNTg3NjExOSwicGF0aCI6Ii8xMzA4MTE3NjUvMzY1NTc0Mjc5LTc5MTEwOTJmLTYyMjEtNDM2NS05M2FmLTVjNGY5NGI4ZmNhNS5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwOVQxMDAxNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MTAwZDE2NmVlNDE5ZGU5ZjVlMmI0ZmNkMGRlYTVlNDE3YmEzZTUxMmRiNTA0MjMwYTM2YmY5NmJiZTE1NWI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.0AGyk9Xn5sTeNzl225NG77C6-SIuNYamMJ4-kPC4cQw)
> 
> And one more problem - t5 gguf doesn't load. [ERROR: Expected all tensors to be on the same device](https://github.com/Juqowel/GPU_For_T5/issues/1#issue-2512755556)

I disagree, Text-encoders first are loaded into the VRAM. I say that because I can see my VRAM usage is up by 10.5GB (t5xxl_fp16.safetensors +clip_l.safetensors = 10GB) first before Forge processes other models. Then the memory management tries to empty the VRAM, then load it again. When I activate your extension, I can see the VRAM gets loaded with the flux model directly in one go.


On a different note, I noticed that the "GPU Weights" feature does not work when using the extension. This results in the VRAM being entirely used When using fp16) and the generation is extremely slow as you can see from the screenshot. With the extension deactivated, the GPU Weights feature works but the issue is the RAM and Disk usage stays at 100% for several minutes, which results in either my machine crashing or generation is successful. When it is successful, the speed is 1.45s/t.
![image](https://github.com/user-attachments/assets/29d68bdb-07b7-45cf-86f6-64e982ecea23)

**Reply by Juqowel** [reply]
> RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0!

That's gguf load logic. Can't do anything with it now.

> I disagree

To be clear. (this is 32GB)
With extension and two gpu:
![ramm](https://github.com/user-attachments/assets/c7cc4ef6-0e7e-4fa0-9a4b-8109f7012b4b)
First - load t5 to RAM -> load t5 to VRAM -> clear RAM
Second - load flux(nf4) to RAM -> load flux(nf4) to VRAM -> clear RAM

With extension and gpu + cpu:
![ram](https://github.com/user-attachments/assets/ecdd3a93-6928-4e11-b0e3-3b0bc37ba613)
First - load t5 to RAM -> stays here
Second - load flux(nf4) to RAM -> load flux(nf4) to VRAM - clear RAM

Better if (without extension):
First - load flux to RAM -> load flux to VRAM - clear RAM
Second - load t5 to RAM -> stays here

> On a different note, I noticed that the "GPU Weights" feature does not work when using the extension.

Yes, it's needs for properly work. 
But you can test without it by remove `memory_management.lowvram_available = False`.

Edit: Anyway, I want to rework it in next month.

**Reply by Iory1998** [reply]
> Better if (without extension):
> First - load flux to RAM -> load flux to VRAM - clear RAM
> Second - load t5 to RAM -> stays here

I totally agree.

> Yes, it's needs for properly work.
> But you can test without it by remove memory_management.lowvram_available = False.
What do you mean?

> Edit: Anyway, I want to rework it in next month.
Thank you and good luck. Do your best to work on the extension this month :P

**Reply by Juqowel** [reply]
> What do you mean?

It's just prevents unnecessary offload. But in your case.. I'm not sure. Is it work without extension faster(s/it)?

Upd: option for control LowVram function. Didn't test, but should work.

**Reply by Iory1998** [reply]
@Juqowel I wrote a post comparing different GGUF quants on reddit and the my surprised, many people didn't know about forcing textencoders into RAM was even possible. I referred them to your extension and it worked well for a few. 
I highly request you to give this extension serious development as it would really help people with low VRAM.

> Upd: option for control LowVram function. Didn't test, but should work.

I'll update the extension and try it again.

**UPDATE**
IT WORLED BEAUTIFULLY. Here is what happens. When I launch Forge for the first time, and activate the extension, I see "T5 on Other Device is enabled for txt2img and cpu" message in console but there is no way to see whether "Low RAM" is activated too. IT doesn't activate even when ticked. I had to deactivate and reactivate the extension for it to work with GPU Weights. And when it did, It gave me the best experience with FP16 so far. My ram usage never exceeded 50% in the process. 
Could you update the extension so Low VRAM is activated by default? Maybe this would solve the issue. Also, is there a way the whole extension is activated by default?
@lllyasviel Maybe you could consider including this extension by in Forge by default.

![image](https://github.com/user-attachments/assets/5a0430b6-7213-4a16-abea-5de0cea5fd3b)

**Reply by Juqowel** [reply]
@Iory1998 Thanks for testing. I don't have time for this.
The update is ready.

**Comment by Arnaud3013** [comment]
Really usefull extensions. On dev nf4 (RTX 4070, setting 9500 max vram for model) Great speedup if no model change between generation.
previously:
```
100%|████████████████████████████████████████████████████████████████████| 5/5 [00:07<00:00,  1.60s/it]
Skipping unconditional conditioning when CFG = 1. Negative Prompts are ignored.██████████| 5/5 [00:07<00:00,  1.29s/it]
[Unload] Trying to free 15315.57 MB for cuda:0 with 0 models keep loaded ... Current free memory is 4472.06 MB ... Unload model KModel Current free memory is 10860.76 MB ... Unload model IntegratedAutoencoderKL Done.
[Memory Management] Target: JointTextEncoder, Free GPU: 11020.64 MB, Model Require: 9641.98 MB, Previously Loaded: 0.00 MB, Inference Require: 2781.00 MB, Remaining: -1402.34 MB, CPU Swap Loaded (blocked method): 2886.75 MB, GPU Loaded: 6899.98 MB
Moving model(s) has taken 3.62 seconds
Distilled CFG Scale: 3.5
[Unload] Trying to free 10901.84 MB for cuda:0 with 0 models keep loaded ... Current free memory is 3613.65 MB ... Unload model JointTextEncoder Done.
[Memory Management] Target: KModel, Free GPU: 11015.63 MB, Model Require: 6246.80 MB, Previously Loaded: 0.00 MB, Inference Require: 2781.00 MB, Remaining: 1987.83 MB, All loaded to GPU.
Moving model(s) has taken 3.24 seconds
100%|████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:07<00:00,  1.47s/it]
```
Now:
```
100%|████████████████████████████████████████████████████████████████████| 5/5 [00:08<00:00,  1.76s/it]
Skipping unconditional conditioning when CFG = 1. Negative Prompts are ignored.██████████| 5/5 [00:08<00:00,  1.30s/it]
[Unload] Trying to free 10901.84 MB for cuda:0 with 0 models keep loaded ... Current free memory is 10352.01 MB ... Current free memory is 10352.01 MB ... Unload model IntegratedAutoencoderKL Done.
[Memory Management] Target: KModel, Free GPU: 10511.88 MB, Model Require: 6246.80 MB, Previously Loaded: 0.00 MB, Inference Require: 2781.00 MB, Remaining: 1484.08 MB, All loaded to GPU.
Moving model(s) has taken 1.41 seconds
100%|████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:07<00:00,  1.47s/it]
```
On dev Q8 gguf (same gpu)
Previously:
```
100%|████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:11<00:00,  2.23s/it]
[Unload] Trying to free 15315.57 MB for cuda:0 with 0 models keep loaded ... Current free memory is 10861.01 MB ... Unload model IntegratedAutoencoderKL Done.
[Memory Management] Target: JointTextEncoder, Free GPU: 11020.89 MB, Model Require: 9641.98 MB, Previously Loaded: 0.00 MB, Inference Require: 2781.00 MB, Remaining: -1402.09 MB, CPU Swap Loaded (blocked method): 2886.75 MB, GPU Loaded: 6899.98 MB
Moving model(s) has taken 1.75 seconds
Distilled CFG Scale: 3.5
[Unload] Trying to free 18536.36 MB for cuda:0 with 0 models keep loaded ... Current free memory is 3613.90 MB ... Unload model JointTextEncoder Done.
[Memory Management] Target: KModel, Free GPU: 11015.88 MB, Model Require: 12119.51 MB, Previously Loaded: 0.00 MB, Inference Require: 2781.00 MB, Remaining: -3884.63 MB, CPU Swap Loaded (blocked method): 5202.00 MB, GPU Loaded: 6917.51 MB
Moving model(s) has taken 4.48 seconds
100%|████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:11<00:00,  2.24s/it]
```
after:
```
100%|████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:11<00:00,  2.23s/it]
Distilled CFG Scale: 3.5
[Unload] Trying to free 18536.36 MB for cuda:0 with 0 models keep loaded ... Current free memory is 10354.01 MB ... Unload model IntegratedAutoencoderKL Current free memory is 10513.88 MB ... Done.
[Memory Management] Target: KModel, Free GPU: 10513.88 MB, Model Require: 12119.51 MB, Previously Loaded: 0.00 MB, Inference Require: 2781.00 MB, Remaining: -4386.63 MB, CPU Swap Loaded (blocked method): 5680.12 MB, GPU Loaded: 6439.38 MB
Moving model(s) has taken 2.62 seconds
100%|████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:11<00:00,  2.24s/it]
```
But it seems to have some issues with memory management. When i change model, like in x/y/z plot and do some genration/tests, my memory explode. Vram constant no issue here, blocked at 9000 on my 4070, but RAM it's another topic. I've set some virtual ram to be sure to handle model in q8_0, I've 32go physical RAM and 60go of virtual RAM on nvme ssd. without extension no issues, sometimes RAM usage go up to 55go but not more, with extension very often it go up to 90go and crash forge.
